### DISCUSSION TITLE
CSS-in-JS: `style` vs. `sx` vs. `__css` prop. What's the difference?

### DISCUSSION THREAD

**Discussion by tannguyencse19** [chakra-ui/chakra-ui#5306] [Q&A] [answered]
I'm not notice them until I care about bundle size. Can anyone tell me (or provide me a link) in-depth what's the difference? 

Having search google but there are too many scattering posts, blogs, answer and still I don't understand.

**Comment by primos63** [comment, marked as answer]
In Chakra styles are applied in precedence order (lowest to highest)

- __css has the lowest precedence and is use by components for initial base styling
- baseStyle is next and is the styling that comes from the theme
- otherProps (textStyle, layerStyle, apply) will be next to be applied once the PR for 3884 is promoted
- styleProps these are the props you add on the component for example `<Box bg='blue.200' color='white' mx={4} />`, `bg`, `color` and `mx` are style props
- sx has the highest precedence and will override any of the other styles.

`style` is the same HTML attribute defined by MDN:
> The style global attribute contains CSS styling declarations to be applied to the element. Note that it is recommended for styles to be defined in a separate file or files. This attribute and the <style> element have mainly the purpose of allowing for quick styling, for example for testing purposes.

There may be a few components which use this, but as a general rule of thumb its use is discouraged.

None of this should have anything to do with bundle size.

**Reply by agupta1989** [reply]
I am developing a new custom component of my own using chakra ui component (e.g. input component). while defining the multi part style, where should I apply the CSS object ? in `__css` prop or `sx` prop ? when I am trying to `__css` prop it doesn't work. it worked only with `sx`. any idea why ? (regarding the first point you mentioned).

Following is my example:
`const comboboxStyle = useMultiStyleConfig('Combobox', {}); `

this doesn't work
```
<Input __css={comboboxStyle.input}/>
```

but this work:
```
<Input sx={comboboxStyle.input}/>
```

**Reply by veloware** [reply]
My (limited) knowledge says this is because the `Input` component itself already have a theme, and will already be using `__css` internally.

I actually have a related question around this, which kind of expands on this question further.

Using the above example, I'm also building my own component library on top of Chakra, if I have an custom component that uses something like `Input`, I can't apply styles using `__css` as noted above, but can use style props or `sx`.  If I use `sx`, then the consumer that uses my component cannot then use `sx` otherwise it overrides the internal use of `sx`, is there a recommendation to overcome this?  note, I may be using sx interally for styles like - 

```
sx={{
    '.parent &:first-child: {
         // ...styles
    }
}}
```

which can't be done using style props, if I understand correctly.  Note also, my styles are actually coming from a theme, so I do something like - 

```
const styles = useStyleConfig('x')

sx={styles}
```

where `styles` contains - 

```
{
    '.parent &:first-child: {
         // ...
    }
}
```

or whatever, from the theme
